packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/AndroidIntentPlugin.java

```diff
@@ -1,161 +1,60 @@
-// found in the LICENSE file.
-
 package io.flutter.plugins.androidintent;
 
-import android.content.ComponentName;
-import android.content.Context;
-import android.content.Intent;
-import android.net.Uri;
-import android.os.Bundle;
-import android.provider.Settings;
-import android.util.Log;
-import io.flutter.plugin.common.MethodCall;
-import io.flutter.plugin.common.MethodChannel;
-import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
-import io.flutter.plugin.common.MethodChannel.Result;
-import io.flutter.plugin.common.PluginRegistry.Registrar;
-import java.util.ArrayList;
-import java.util.Map;
-
-/** AndroidIntentPlugin */
-@SuppressWarnings("unchecked")
-public class AndroidIntentPlugin implements MethodCallHandler {
-  private static final String TAG = AndroidIntentPlugin.class.getCanonicalName();
-  private final Registrar mRegistrar;
-
-  /** Plugin registration. */
-  public static void registerWith(Registrar registrar) {
-    final MethodChannel channel =
-        new MethodChannel(registrar.messenger(), "plugins.flutter.io/android_intent");
-    channel.setMethodCallHandler(new AndroidIntentPlugin(registrar));
+import androidx.annotation.NonNull;
+import io.flutter.embedding.engine.plugins.FlutterPlugin;
+import io.flutter.embedding.engine.plugins.activity.ActivityAware;
+import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
+
+/**
+ * Plugin implementation that uses the new {@code io.flutter.embedding} package.
+ *
+ * <p>Instantiate this in an add to app scenario to gracefully handle activity and context changes.
+ */
+public final class AndroidIntentPlugin implements FlutterPlugin, ActivityAware {
+  private final IntentSender sender;
+  private final MethodCallHandlerImpl impl;
+
+  /**
+   * Initialize this within the {@code #configureFlutterEngine} of a Flutter activity or fragment.
+   *
+   * <p>See {@code io.flutter.plugins.androidintentexample.MainActivity} for an example.
+   */
+  public AndroidIntentPlugin() {
+    sender = new IntentSender(/*activity=*/ null, /*context=*/ null);
+    impl = new MethodCallHandlerImpl(sender);
   }
 
-  private AndroidIntentPlugin(Registrar registrar) {
-    this.mRegistrar = registrar;
-  }
-
-  private String convertAction(String action) {
-    switch (action) {
-      case "action_view":
-        return Intent.ACTION_VIEW;
-      case "action_voice":
-        return Intent.ACTION_VOICE_COMMAND;
-      case "settings":
-        return Settings.ACTION_SETTINGS;
-      case "action_location_source_settings":
-        return Settings.ACTION_LOCATION_SOURCE_SETTINGS;
-      case "action_application_details_settings":
-        return Settings.ACTION_APPLICATION_DETAILS_SETTINGS;
-      default:
-        return action;
-    }
+  @Override
+  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
+    sender.setApplicationContext(binding.getApplicationContext());
+    sender.setActivity(null);
```

Comment:
Would it make more senes to set activity to null in the appropriate `ActivityAware` methods? The execution of this method doesn't really imply anything about the Activity...

Reply:
Would a user potentially be changing the plugin's engine, but expecting it to hold on to an existing activity reference? I was assuming that if the engine changes, the activity itself is also invalid. I also figured that the Activity callbacks should be called anyway so I am updating the Activity there too, but I wanted to add an extra check here just to guarantee that they never got out of state based on some callback issue.

Reply:
You can add an extra call, but if that extra call ever has an effect, it indicates a bug in the embedding. The Activity calls are supposed to be strictly between engine calls:

attach engine
attach activity
detach activity
detach engine